controllers/applicationvolumereplication_controller.go

```diff
@@ -195,295 +191,471 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+	}
+
+	// Restore PVs to the new home cluster
+	return a.restore(newHomeCluster)
+}
 
-	pvMWName := fmt.Sprintf(ManifestWorkNameFormat, subscription.Name, subscription.Namespace, MWTypePV)
+func (a *AVRInstance) cleanup() error {
+	// first delete the MW for PVs
+	pvMWName := BuildManifestWorkName(a.instance.Name, a.instance.Namespace, MWTypePV)
 
-	err := a.deleteExistingManifestWork(subscription, pvMWName)
+	err := a.deleteExistingManifestWork(pvMWName)
 	if err != nil {
-		a.log.Error(err, "Failed to delete existing PV manifestwork for subscription", "name", subscription.Name)
+		a.log.Error(err, "Failed to delete existing PV manifestwork")
 
-		return !unpause
+		return err
 	}
 
-	vrgMWName := fmt.Sprintf(ManifestWorkNameFormat, subscription.Name, subscription.Namespace, MWTypeVRG)
+	// Next, Update VRG to secondary
+	vrgMWName := BuildManifestWorkName(a.instance.Name, a.instance.Namespace, MWTypeVRG)
 
-	err = a.deleteExistingManifestWork(subscription, vrgMWName)
+	err = a.updateVRGToSecondary(vrgMWName)
 	if err != nil {
-		a.log.Error(err, "Failed to delete existing VRG manifestwork for subscription", "name", subscription.Name)
+		a.log.Error(err, "Failed to update existing VRG manifestwork")
 
-		return !unpause
+		return err
 	}
 
-	err = a.restorePVFromBackup(subscription, newHomeCluster)
+	return nil
+}
+
+func (a *AVRInstance) restore(newHomeCluster string) error {
+	// Restore from PV backup location
+	err := a.restorePVFromBackup(newHomeCluster)
 	if err != nil {
-		a.log.Error(err, "Failed to restore PVs from backup for subscription", "name", subscription.Name)
+		a.log.Error(err, "Failed to restore PVs from backup")
 
-		return !unpause
+		return err
 	}
 
-	a.moveToNextDRState(subscription.Name)
-
-	return unpause
+	return nil
 }
 
-func (a *AVRInstance) vrgManifestWorkAlreadyExists(
-	subscription *subv1.Subscription) (bool, error) {
-	if a.instance.Status.Decisions == nil {
-		return false, nil
+func (a *AVRInstance) createVRGManifest() bool {
+	a.log.Info("Processing deployment", "Last State", a.instance.Status.LastKnownDRState)
+
+	const requeue = true
+
+	homeCluster, peerCluster, err := a.extractHomeClusterAndPeerCluster()
+	if err != nil {
+		a.log.Info(fmt.Sprintf("Unable to select placement decision (%v)", err))
+
+		return requeue
 	}
 
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		// Skip this subscription if a manifestwork already exist for it
-		mwName := BuildManifestWorkName(subscription.Name, subscription.Namespace, MWTypeVRG)
+	mw, err := a.vrgManifestWorkAlreadyExists(homeCluster)
+	if err != nil {
+		return requeue
+	}
 
-		mw, err := a.findManifestWork(mwName, d.HomeCluster)
+	if mw != nil {
+		primary, err := a.isVRGPrimary(mw)
 		if err != nil {
-			a.log.Error(err, "findManifestWork()", "name", subscription.Name)
+			a.log.Error(err, "Failed to check whether the VRG is primary or not", "mw", mw.Name)
 
-			return false, err
+			return requeue
 		}
+		if primary {
```

Comment:
If not primary it will create VRG again below? possibly needs a default return requeue in this block?

Reply:
No that's correct.  We want to create a new VRG as primary. The VRG that we found and it is secondary might be waiting for deletion.  IOW, we might be waiting for the ManagedClusterView to tell us that it is clear to delete it.

Reply:
Ack! I mis-read `isVRGPrimary` as checking state and not spec.